### PR TITLE
Offloading 3/3: enable rematerialization using xla flags

### PR DESCRIPTION
**Pull request by zhenying-liu**
This enables activation offloading automatically on GPU through HLO Rematerialization.
There is one xla flag needed to trigger rematerialization offloading:
1. A bool to enable host memory offloading. It's false by default to disable the offloading.

The other information is achieved by querying the GPU device:
2. Memory bandwidth in bytes per second.
3. The number of floating point operations on a device.

### CODE REVIEW

**Comment by zhenying-liu** [comment]
The memory bandwidth and flops numbers of Grace Hopper are taken from this doc: https://resources.nvidia.com/en-us-grace-cpu/grace-hopper-superchip. It disables HLO rematerialization offloading for GPUs by default, and can enable host memory offloading via those three XLA flags. Since hlo_rematerialization_test.cc already has the unit tests for enabling offloading, we didn't provide the unit test here.

File: xla/service/gpu/gpu_compiler.cc [hunk 1]
```diff
@@ -2194,26 +2194,69 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
...
+      }
+    } else {
+      // Enable host memory offloading in Rematerialization pass
+      HloCostAnalysis::Options hlo_cost_analysis_options;
+      hlo_cost_analysis_options.shape_size = ShapeSizeBytesFunction();
+      hlo_cost_analysis_options.set_flops_per_second(flops_per_sec);
+      hlo_cost_analysis_options.set_transcendentals_per_second(flops_per_sec);
+      HloCostAnalysis hlo_cost_analysis(hlo_cost_analysis_options);
+      HloRematerialization::RematerializationModeConfig
+          rematerialization_mode_config(/*recompute=*/false, /*compress=*/false,
```
Comment (SandSnip3r) [inline xla/service/gpu/gpu_compiler.cc:2238]: Why do you disable other forms of rematerialization when the host offloading strategy is enabled?

File: xla/service/gpu/gpu_compiler.cc [hunk 2]
```diff
@@ -2194,26 +2194,69 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
...
+    float offloading_bandwidth =
+        module->config()
+            .debug_options()
+            .xla_gpu_offloading_bandwidth_bytes_per_sec();
+    float flops_per_sec =
+        module->config().debug_options().xla_gpu_offloading_flops_per_sec();
+    int64_t host_memory_space_color =
+        static_cast<int64_t>(se::MemoryType::kHost);
+
+    if (offloading_memory_limit_gb == 0) {
```
Comment (SandSnip3r) [inline xla/service/gpu/gpu_compiler.cc:2209]: There is a lot of duplicate code here. You should have one bit of code which builds the HloRematerialization::Options (based on the flags, etc.), then have just a single place where the passes are added to the pipeline.

File: xla/service/gpu/gpu_compiler.cc [hunk 3]
```diff
@@ -2194,26 +2194,69 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
...
-    if (changed) {
-      VLOG(1) << "HloRematerialization saved "
-              << sizes.before_bytes - sizes.after_bytes << " bytes";
+    int64_t offloading_memory_limit_gb =
+        module->config().debug_options().xla_gpu_offloading_memory_limit_gb() *
+        1024 * 1024 * 1024;
+    float offloading_bandwidth =
+        module->config()
+            .debug_options()
+            .xla_gpu_offloading_bandwidth_bytes_per_sec();
```
Comment (SandSnip3r) [inline xla/service/gpu/gpu_compiler.cc:2203]: Does it really make sense to take this from a flag? Wouldn't this be queryable based on the current GPU?

**Review by SandSnip3r — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gpu_compiler.cc (hunk 1 above)
Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Sure, recompute/compress remain true whether host offloading is on.

File: xla/service/gpu/gpu_compiler.cc (hunk 2 above)
Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Thanks for pointing this out. It's changed accordingly.

File: xla/service/gpu/gpu_compiler.cc (hunk 3 above)
Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Right. Removed this flag by querying the current GPU.

File: xla/service/gpu/gpu_compiler.cc [hunk 4]
```diff
@@ -2189,66 +2189,44 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
     int64_t offloading_memory_limit_gb =
         module->config().debug_options().xla_gpu_offloading_memory_limit_gb() *
         1024 * 1024 * 1024;
-    float offloading_bandwidth =
-        module->config()
-            .debug_options()
-            .xla_gpu_offloading_bandwidth_bytes_per_sec();
```
Comment (nouiz) [inline xla/service/gpu/gpu_compiler.cc]: Can you also infer flops_per_sec ?
gpu_device_info has fpus_per_core and other function like clock_rate_ghz, that maybe allow you to infer it.

Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Let me figure it out.

Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Modified and verified with a V100 locally.

File: xla/debug_options_flags.cc [hunk 5]
```diff
@@ -1681,6 +1682,13 @@ void MakeDebugOptionsFlags(std::vector<tsl::Flag>* flag_list,
       bool_setter_for(&DebugOptions::set_xla_gpu_use_memcpy_local_p2p),
       debug_options->xla_gpu_use_memcpy_local_p2p(),
       "Whether to use memcpy for local p2p communication."));
+  flag_list->push_back(tsl::Flag(
+      "xla_gpu_offloading_memory_limit_gb",
+      int64_setter_for(&DebugOptions::set_xla_gpu_offloading_memory_limit_gb),
+      debug_options->xla_gpu_offloading_memory_limit_gb(),
+      "GPU memory usage upper limit in GB to trigger host memory offloading."
```
Comment (nouiz) [inline xla/debug_options_flags.cc:1689]: The usage is strange.
Right now, there is already a memory setting for that pass via scheduler_mem_limit.
There is already some flags to adjust it.
What this new version will provide?
Why not make this a bool that default to false? This way, user don't need to find the default GPU memory threshold that is set to use it. So it  would be easier to use.

File: xla/service/gpu/gpu_compiler.cc [hunk 6]
```diff
@@ -2210,22 +2210,44 @@ absl::Status GpuCompiler::RunPostSchedulingPipelines(
...
-                                      /*host_offload=*/false);
+                                      /*host_offload=*/enable_offloading);
+    HloCostAnalysis::Options hlo_cost_analysis_options;
+    hlo_cost_analysis_options.shape_size = ShapeSizeBytesFunction();
+    if (enable_offloading) {
+      hlo_cost_analysis_options.set_flops_per_second(flops_per_sec);
+      hlo_cost_analysis_options.set_transcendentals_per_second(flops_per_sec);
+    }
+    HloCostAnalysis hlo_cost_analysis(hlo_cost_analysis_options);
+    HloRematerialization::HostMemoryOffloadConfig offloading_config(
```
Comment (SandSnip3r) [inline xla/service/gpu/gpu_compiler.cc:2233]: If offloading is not enabled, you should not be constructing a host memory offloading config nor passing it to the HloRematerialization options.

I think it is completely ok if you do, but it's an absolute waste and maybe confusing to someone else.

File: xla/debug_options_flags.cc (hunk 5 above)
Comment (zhenying-liu) [reply xla/debug_options_flags.cc]: OK, I changed it to bool. I thought this memory limit could be smaller than scheduler_mem_limit in order to kick in offloading earlier. I agree it's better to have one flag to control the memory limit.

File: xla/service/gpu/gpu_compiler.cc (hunk 6 above)
Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc:2239]: > If offloading is not enabled, you should not be constructing a host memory offloading config nor passing it to the HloRematerialization options.
How to pass a std::nullopt and offloading_config in an implicit if statement to construct a remat pass? I couldn't solve the type mismatch. It'll be nice to do it.

File: xla/debug_options_flags.cc (hunk 5 above)
Comment (zhenying-liu) [reply xla/debug_options_flags.cc]: I have to change it back to int64_t. The activation offloading memory limit is much more smaller than scheduler_mem_limit. The former could be just few tensors for offloading candidates in a module, while the latter is the whole GPU memory used for JAX. If we use the scheduler_mem_limit for offloading, we can not enable offloading.

File: xla/service/gpu/gpu_compiler.cc (hunk 6 above)
Comment (SandSnip3r) [reply xla/service/gpu/gpu_compiler.cc:2236]:
```
    HloCostAnalysis::Options hlo_cost_analysis_options;
    hlo_cost_analysis_options.shape_size = ShapeSizeBytesFunction();
    std::optional<HloRematerialization::HostMemoryOffloadConfig> offloading_config;
    if (enable_offloading) {
      hlo_cost_analysis_options.set_flops_per_second(flops_per_sec);
      hlo_cost_analysis_options.set_transcendentals_per_second(flops_per_sec);
      offloading_config.emplace(/*host_memory_space=*/host_memory_space_color,
                                /*bandwidth_to_host_bytes_per_second=*/
                                gpu_device_info.memory_bandwidth(),
                                /*bandwidth_from_host_bytes_per_second=*/
                                gpu_device_info.memory_bandwidth());
    }
    HloCostAnalysis hlo_cost_analysis(hlo_cost_analysis_options);
    ```
    Then pass offloading_config unconditionally to the options struct

Comment (zhenying-liu) [reply xla/service/gpu/gpu_compiler.cc]: Thanks so much for showing me how to write such a code. I made a similar version and verified.

**Review by SandSnip3r — APPROVED** [review]

**Pull request closed without merging**